### DISCUSSION TITLE
Contribute custom actions from custom extension of Code for IBMi

### DISCUSSION THREAD

**Discussion by m-tyler** [codefori/vscode-ibmi#1963] [General]
I have a custom extension of Code for IBMi that I am building and I want to have actions that other developers in my company can get when the extension is installed/activated. 

Is there a simple way to contribute custom actions to the actions list?

Thanks, Matt

**Comment by worksofliam** [comment]
I once did come up with an idea for an API where other extensions could register their own Actions. Here is the closed PR for it.

https://github.com/codefori/vscode-ibmi/pull/760/files

If this is a requirement for you, we can open it back up and implement it into our new API code.

@codefori/core

**Comment by sebjulliand** [comment]
Since you're building an extension, instead of adding more `Code for IBM i` custom actions, wouldn't it be better to have your extension contribute regular VSCode actions? 🤔
They could run on any item from any Code for IBM i view and they would integrate nicely with VSCode UI 😄

**Comment by m-tyler** [comment]
@worksofliam :As  a requirement, no its not for me to finish my extension and its not a priority for me. 

@sebjulliand: I don't know how to really answer this.  Best answer may be I don't know.   A few of  actions I wanted to have other programmers use are not wrapped into VS Code commands, yet, if at all.  Also, they need member variable information to operate.

**Reply by sebjulliand** [reply]
Can you provide an example of an action you'd like to add (needing member info, for example)? I can give you a few pointers if you want.

**Reply by m-tyler** [reply]
This action command right now cannot be wrapped in a VS Code command. 
```
{
      "type": "member",
      "extensions": [
        "GLOBAL"
      ],
      "environment": "ile",
      "name": "🚸HAWKEYE: Print RPG Report spacing",
      "command": "?PRTRPGPRT SRCFILE(&OPENLIB/&OPENSPF) SRCMBR(&OPENMBR) SUMMARY(*NO) OUTPUT(*PRINT) PRTFILE(qsysprt)",
      "refresh": "no",
      "runOnProtected": true
    },
```
I used this on old RPGLE programs with internal O-specs that I cannot easily convert to external PRTF. 

Selecting to run the action command.
![image](https://github.com/codefori/vscode-ibmi/assets/1353665/0f965fdc-7aa0-4e85-a4a1-54c313c5bb21)

SImple prompt of action.
![image](https://github.com/codefori/vscode-ibmi/assets/1353665/d1c08795-5307-4845-8be1-4add39f1a162)

Results always go to QSYSPRT never to the OUTPUT channel so I don't have a way to scoop up the results, yet.  
![image](https://github.com/codefori/vscode-ibmi/assets/1353665/f5a066c6-e332-4472-bed0-7d4cccf7a5ab)

Print report showing me the report layout. (printer file named is an generic name allowing for OUTQ direction.)
![image](https://github.com/codefori/vscode-ibmi/assets/1353665/26185d1b-ae56-43b3-a67e-a8b2220eaccd)

**Reply by sebjulliand** [reply]
I'll get back to you tomorrow if you wish, with an example of how you can call this program from a VSCode command and maybe then call the `SPOOLED_FILE_DATA` SQL service to get that output.

**Reply by m-tyler** [reply]
Thanks, but its no hurry.   I have a few other things to complete on this extension before its ready.  
 
I did not write this with a VS Code command mostly because of the output generated that I want is only offered as a SPLF.   But I have several existing commands I wrote for this extension that I can pattern and do that.   The IBM i command (HAWKEYE/PRTRPGPRT) shown actually has an OUPUT(*OUTFILE) option but its does not produce the layout that way.   

I have found that I had issues using `SPOOLED_FILE_DATA` with reports that are coded as *DBCS but are not actually produced with that content.   I have a function for another project for user reports that downloads SPLF content and can open it up into a edit tab.  I am looking to go this route in the future but in the meantime I was looking to see if I could provide my other developers with actions for features I have not written a command over.

**Reply by sebjulliand** [reply]
No worries, it didn't take long to put together. It's a rough example of how you do what you described but using the exported API instead of commands.

You can find the example here: https://github.com/sebjulliand/command-extension

It's a simple extension that adds a submenu to the member items with one command:
![image](https://github.com/codefori/vscode-ibmi/assets/11096890/88dc9dab-af2b-43b0-9372-135c8d6e0ef2)

That's just to give you an idea of what's possible and get you inspired 😄

**Reply by m-tyler** [reply]
Thanks I'll look it over and morph it into my extension.
